recirq/quantum_chess/circuit_transformer_test.py

```diff
@@ -29,7 +29,60 @@
 c3 = cirq.NamedQubit('c3')
 d1 = cirq.NamedQubit('d1')
 
+@pytest.mark.parametrize('device',
```

Comment:
Is there a particular test case where map_helper failed?  We should add it as a regression test.  It will also make it more clear why we need to adjust the search function.

Reply:
Hi Doug, the last test in this file covers the situation corresponding to the map_helper() bug. That test will fail the bug is revealed if the fix is not been applied.

Reply:
the last test that you added? or the last test that was already in the file?

I'm just curious why circuit_transformer_test was not failing previously / is not failing without changes from this pull request -- if it were it should have stopped us from merging previous pull requests?

Reply:
The last test that was already in the file. Its debug output for device0 is:

**Before change:**

a1 -> (6, 3)
a2 -> (6, 4)
a3 -> (6, 5)
a4 -> (7, 5)
Not adjacent a1 and a4
Warning: could not map all qubits!
a3 -> (7, 4)
a4 -> (7, 5)
Not adjacent a1 and a4
a4 -> (7, 3)
a3 -> (7, 4)
a2 -> (6, 4)
b3 -> (6, 5)
a2 -> (6, 4)
a1 -> (6, 3)
b1 -> (6, 2)
c1 -> (6, 1)
d1 -> (5, 1)
Not adjacent a1 and d1
Warning: could not map all qubits!
c1 -> (7, 2)
Cannot fit adjacent nodes into c1
c1 -> (5, 2)
d1 -> (5, 3)
c1 -> (5, 2)
b1 -> (6, 2)
c2 -> (6, 1)
b1 -> (6, 2)
a1 -> (6, 3)

So it seems that even though before changing it starts with a wrong grid qubit, it ends up getting an accepted solution. But **after the top 2 DFS fixes** the problem shows up:

        for node_to_try in valid_adjacent_qubits:
            # Add proposed qubit to the mapping
            # and remove it from available qubits
            mapping[node_to_map] = node_to_try
>           available_qubits.remove(node_to_try)
E           KeyError: cirq.GridQubit(6, 3)

circuit_transformer.py:204: KeyError
---------------------------------------------------------- Captured stdout call -----------------------------------------------------------
a1 -> (5, 3)
a2 -> (5, 4)
Cannot fit adjacent nodes into a2
a2 -> (5, 2)
a3 -> (5, 1)
a4 -> (5, 0)
Not adjacent a1 and a4
a4 -> (6, 1)
Not adjacent a1 and a4
a4 -> (4, 1)
Not adjacent a1 and a4
Warning: could not map all qubits!
a3 -> (6, 2)
a4 -> (6, 3)
a3 -> (6, 2)
a2 -> (5, 2)
b3 -> (5, 1)
a2 -> (5, 2)
a1 -> (5, 3)
b1 -> (5, 4)
Cannot fit adjacent nodes into b1
b1 -> (4, 3)
Cannot fit adjacent nodes into b1
Warning: could not map all qubits!

**After applying all 3 fixes:**

a1 -> (5, 3)
a2 -> (5, 4)
Cannot fit adjacent nodes into a2
a2 -> (5, 2)
a3 -> (5, 1)
a4 -> (5, 0)
Not adjacent a1 and a4
a4 -> (6, 1)
Not adjacent a1 and a4
a4 -> (4, 1)
Not adjacent a1 and a4
Warning: could not map all qubits!
a3 -> (6, 2)
a4 -> (6, 3)
a3 -> (6, 2)
a2 -> (5, 2)
b3 -> (5, 1)
a2 -> (5, 2)
a1 -> (5, 3)
b1 -> (5, 4)
Cannot fit adjacent nodes into b1
b1 -> (4, 3)
Cannot fit adjacent nodes into b1
Warning: could not map all qubits!
a2 -> (6, 3)
a3 -> (6, 4)
a4 -> (6, 5)
Not adjacent a1 and a4
a4 -> (7, 4)
Not adjacent a1 and a4
a4 -> (5, 4)
a3 -> (6, 4)
a2 -> (6, 3)
b3 -> (6, 2)
a2 -> (6, 3)
a1 -> (5, 3)
b1 -> (5, 2)
c1 -> (5, 1)
d1 -> (5, 0)
Not adjacent a1 and d1
d1 -> (6, 1)
Not adjacent a1 and d1
d1 -> (4, 1)
Not adjacent a1 and d1
Warning: could not map all qubits!
c1 -> (4, 2)
d1 -> (4, 3)
c1 -> (4, 2)
b1 -> (5, 2)
c2 -> (5, 1)
b1 -> (5, 2)
a1 -> (5, 3)

Reply:
I see. So to Doug's point, can we come up with a test case that would have failed before the fixes (and without getting lucky) but passes now? That will help prevent future regressions if this code is refactored again in the future.